plugins/saveactions.c

```diff
@@ -304,6 +314,323 @@ static void instantsave_document_new_cb(GObject *obj, GeanyDocument *doc, gpoint
+	msg2 = _("Do you want to save it before closing?");
+
+	response = run_unsaved_dialog_for_persistent_temp_files_tab_closing(msg, msg2);
+	g_free(msg);
+
+	locale_file_path = g_strdup(doc->real_path);
+
+	switch (response)
+	{
+		case GTK_RESPONSE_YES:
+			old_file_path = g_strdup(doc->file_name);
+
+			if (dialogs_show_save_as())
+			{
+				new_file_path = doc->file_name;
+
+				if (! g_str_equal(old_file_path, new_file_path))
+				{
+				 	/* remove temp file if it was saved as some other file */
+					g_remove(locale_file_path);
+				}
+			}
+			else
+			{
+				plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			}
+
+			g_free(old_file_path);
+			break;
+
+		case GTK_RESPONSE_NO:
+			g_remove(locale_file_path);
+
+			ui_set_statusbar(TRUE, _("Temp file %s was deleted"), short_filename);
+			break;
+
+		case GTK_RESPONSE_CANCEL:
+		default:
+			plugin_idle_add(geany_plugin, open_document_once_idle, g_strdup(locale_file_path));
+			break;
+	}
+
+	g_free(locale_file_path);
+}
+
+
+static void persistent_temp_files_document_close_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	gchar *short_filename;
+
+	if (enable_persistent_temp_files && doc->real_path != NULL
+		&& is_temp_saved_file_doc(doc) && ! geany_is_closing_all_tabs())
+	{
+		short_filename = document_get_basename_for_display(doc, -1);
+
+		if (! document_is_empty(doc))
+		{
+			show_unsaved_dialog_for_persistent_temp_files_tab_closing(
+				doc,
+				short_filename
+			);
+		}
+		else
+		{
+			g_remove(doc->real_path);
+
+			ui_set_statusbar(TRUE, _("Empty temp file %s was deleted"), short_filename);
+		}
+
+		g_free(short_filename);
+	}
+}
+
+
+static void persistent_temp_files_document_before_save_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	if (enable_persistent_temp_files)
+	{
+		gchar *old_file_path, *new_file_path, *old_file_name;
+				GtkWidget *page_label = NULL;
+
+		get_current_tab_label(&page_label);
+
+		old_file_path = gtk_widget_get_tooltip_text(page_label);
+		new_file_path = DOC_FILENAME(doc);
+
+		if (old_file_path == NULL)
+		{
+			ui_set_statusbar(TRUE, _("plugin error: failed to delete initial temp file "
+				"('failed to get notebook tab label text')"));
+
+			return;
+		}
+
+		old_file_name = g_path_get_basename(old_file_path);
+
+		if (is_temp_saved_file(old_file_name) && ! g_str_equal(old_file_path, new_file_path))
+		{
+			/* we have to store old/new filename pair inside document data to be able to somehow
+			pass it to document-save callback that is called directly after this one */
+			plugin_set_document_data_full(geany_plugin, doc, new_file_path, g_strdup(old_file_path), g_free);
+		}
+
+		g_free(old_file_name);
+	}
+}
+
+
+static void persistent_temp_files_document_save_cb(GObject *obj, GeanyDocument *doc, gpointer user_data)
+{
+	gchar *new_file_path, *old_file_path;
+
+	new_file_path = DOC_FILENAME(doc);
+	old_file_path = plugin_get_document_data(geany_plugin, doc, new_file_path);
```

Comment:
Instead of indexing it by `new_file_path` you could use just some static string like `"old_path"` - the extra data is attached to a document so there's no risk you'll get some other's document path.

Reply:
done